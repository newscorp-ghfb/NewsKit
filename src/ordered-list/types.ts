import {MQ} from '../utils/style';
import {LogicalProps} from '../utils/logical-properties';

export interface OrderedListProps {
  children: React.ReactNode;
  overrides?: {
    spaceInline?: MQ<string>; // LOGICAL_PROPS_TO_DO: Used as the gap between items. Should be renamed.
    content?: {
      stylePreset?: MQ<string>;
      typographyPreset?: MQ<string>;
    };
    counter?: {
      stylePreset?: MQ<string>;
      typographyPreset?: MQ<string>;
      minWidth?: MQ<string>;
    };
  } & LogicalProps;
}
