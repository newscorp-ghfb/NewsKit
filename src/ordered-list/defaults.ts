export default {
  orderedList: {
    spaceInline: 'space040',
    content: {
      stylePreset: 'inkBase',
      typographyPreset: 'editorialParagraph010',
    },
    counter: {
      stylePreset: 'inkBase',
      typographyPreset: 'editorialParagraph010',
      minWidth: 'sizing050',
    },
  },
};
