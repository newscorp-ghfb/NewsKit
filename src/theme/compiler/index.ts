/* eslint-disable @typescript-eslint/no-explicit-any */
import {Theme, UncompiledTheme} from '../types';
import {ThemeCompilerOptions} from './types';
import {generateInkStylePresets} from './ink-style-preset-generator';
import {recurseUnknown} from '../../utils/recurse-unknown';

export const compileTheme = (
  theme: UncompiledTheme | Theme,
  options: ThemeCompilerOptions = {},
): Theme => {
  if (!theme || theme.compiled === true) {
    return theme;
  }

  // eslint-disable-next-line no-console
  const errorLogger = options.errorLogger || console.error.bind(console);

  const uncompiledTheme = {
    ...theme,
    stylePresets: {
      ...theme.stylePresets,
      ...generateInkStylePresets(theme),
    },
  };

  const {icons = {}, ...filteredTheme} = uncompiledTheme;
  return {
    ...(recurseUnknown(uncompiledTheme, filteredTheme, errorLogger) as any),
    icons,
    compiled: true,
  };
};
