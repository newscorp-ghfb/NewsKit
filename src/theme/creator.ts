import {deepMerge} from '../utils/deep-merge';

import {
  BreakpointKeys,
  Breakpoints,
  ThemeBase,
  ThemeLoggerFunction,
  ThemeOverrides,
  UncompiledTheme,
} from './types';

import * as foundations from './foundations';
import * as presets from './presets';

import {get} from '../utils/get';
import {mergeBreakpointObject} from '../utils/merge-breakpoint-object';
import {FontConfig} from './foundations/fonts';

export interface CreateThemeArgs {
  name?: string;
  baseTheme?: UncompiledTheme;
  overrides?: ThemeOverrides;
  checkOverrides?: boolean;
  warningLogger?: ThemeLoggerFunction;
}

const deepDuplicationChecker = (
  warningLogger: ThemeLoggerFunction,
  baseTheme: ThemeBase,
  overrides: ThemeOverrides,
) => {
  const recurse = (obj: object, keys?: string) => {
    Object.entries(obj).forEach(([key, value]) => {
      const newKey = keys ? [keys, key].join('.') : key;
      if (typeof value === 'object') {
        recurse(value, newKey);
      } else if (get(baseTheme, newKey) === value) {
        warningLogger(
          `Override at path: '${newKey}' has the same value as base theme: '${value}'.`,
        );
      }
    });
  };
  recurse(overrides);
};

const isFontConfig = (x: string | FontConfig): x is FontConfig =>
  (x as FontConfig).fontFamily !== undefined;

// Font family info should not be merged. E.g.:
// - Base theme: { ...fontFamily010: {fontFamily: "Font A", fontMetrics: {...fontAMetrics} }
// - Overrides: { ...fontFamily010: {fontFamily: "Font B"} }
// In this example, we should remove fontFamily010 from the base theme, otherwise
// the merged object will be invalid:
// - { ...fontFamily010: {fontFamily: "Font B", fontMetrics: {...fontAMetrics} }
// This function removes font families from a theme if they are defined in an
// overrides object (which can be another theme). It should be called before merging
// the theme with the overrides.
const removeFontFamilies = <T extends ThemeBase>(
  theme: T,
  overrideFonts: T['fonts'],
): T => ({
  ...theme,
  fonts: theme.fonts
    ? Object.entries(theme.fonts).reduce<T['fonts']>(
        (prev, [k, v]) => ({
          ...prev,
          ...(isFontConfig(v) && overrideFonts[k] ? {} : {[k]: v}),
        }),
        {},
      )
    : undefined,
});

export const createTheme = ({
  name = 'unnamed-newskit-theme',
  baseTheme,
  overrides = {},
  checkOverrides,
  // eslint-disable-next-line no-console
  warningLogger = console.warn.bind(console),
}: CreateThemeArgs): UncompiledTheme => {
  if (baseTheme && baseTheme.compiled) {
    throw new Error(
      'createTheme received a compiled baseTheme. Base themes must be uncompiled.',
    );
  }

  // Don't take font family info from NKLight if defined in the base theme.
  const newskitLight = removeFontFamilies<ThemeBase>(
    {
      ...foundations,
      ...presets,
      componentDefaults: {},
      icons: {},
    },
    baseTheme?.fonts || {},
  );

  if (checkOverrides) {
    deepDuplicationChecker(warningLogger, baseTheme || newskitLight, overrides);
  }

  /* istanbul ignore next */
  const breakpointsKeys = (overrides?.breakpoints
    ? /* istanbul ignore next */
      overrides?.breakpoints
    : newskitLight.breakpoints) as Breakpoints;

  return deepMerge(
    mergeBreakpointObject(Object.keys(breakpointsKeys) as BreakpointKeys[]),
    newskitLight,
    // Don't take font family info from the base theme if defined in overrides.
    baseTheme
      ? removeFontFamilies<UncompiledTheme>(baseTheme, overrides?.fonts || {})
      : baseTheme,
    overrides,
    {name, themeVersion: 1},
  );
};
