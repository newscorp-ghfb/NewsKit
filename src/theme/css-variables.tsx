import {css} from '@emotion/react';
import styled from '@emotion/styled';
import {hasOwnProperty} from '../utils/has-own-property';
import {Theme, ThemeBase} from './types';

type FoundationKey = keyof Omit<
  ThemeBase,
  | 'breakpoints'
  | 'stylePresets'
  | 'typographyPresets'
  | 'transitionPresets'
  | 'componentDefaults'
  | 'icons'
>;

const foundationsList: FoundationKey[] = [
  'borders',
  'colors',
  'overlays',
  'motions',
  'shadows',
  'sizing',
  'spacePresets',
  'fonts',
];

export type ThemeDiff = Partial<Pick<ThemeBase, FoundationKey>>;

type TokenValue = string | number | {fontFamily: string};

const hasSameValue = (
  tokenName: string,
  tokenValueA: TokenValue,
  tokenValueB: TokenValue,
) => {
  if (
    tokenName.startsWith('fontFamily') &&
    typeof tokenValueA === 'object' &&
    typeof tokenValueB === 'object'
  ) {
    return tokenValueA.fontFamily === tokenValueB.fontFamily;
  }
  return tokenValueA === tokenValueB;
};

export const themeDiff = (
  parentTheme: Partial<Theme>,
  childTheme: Partial<Theme>,
): ThemeDiff => {
  const diff = {} as ThemeDiff;

  foundationsList.forEach(foundationKey => {
    const parentFoundation = parentTheme[foundationKey] || {};
    const childFoundation = childTheme[foundationKey] || {};

    Object.entries(childFoundation).forEach(([tokenName, tokenValue]) => {
      if (
        foundationKey &&
        tokenName &&
        (!parentFoundation[tokenName] || // does not exist in parent theme, so is new token
          !hasSameValue(tokenName, parentFoundation[tokenName], tokenValue)) // token value are different
      ) {
        if (!hasOwnProperty(diff, foundationKey)) {
          diff[foundationKey] = {};
        }

        // @ts-ignore
        diff[foundationKey][tokenName] = tokenValue;
      }
    });
  });

  return diff;
};

export const generateCSSVars = (theme: Partial<Theme>) => {
  const cssValue = foundationsList
    .map(themeKey => {
      const tokensObject = theme[themeKey];
      /* istanbul ignore if */
      if (typeof tokensObject !== 'object') {
        return '';
      }
      const tokensNames = Object.keys(tokensObject);
      const prefix = themeKey === 'colors' ? 'color-' : '';
      return tokensNames
        .map(tokenName => {
          if (
            tokenName.startsWith('fontFamily') &&
            typeof tokensObject[tokenName] === 'object' &&
            tokensObject[tokenName].fontFamily
          ) {
            return `--${prefix}${tokenName}: ${tokensObject[tokenName].fontFamily};`;
          }
          return `--${prefix}${tokenName}: ${tokensObject[tokenName]};`;
        })
        .join('');
    })
    .join('');

  return css`
    ${cssValue}
  `;
};

export const CssVariablesContainer = styled.div<{diff: ThemeDiff}>`
  ${({diff}) => generateCSSVars(diff)}
`;
