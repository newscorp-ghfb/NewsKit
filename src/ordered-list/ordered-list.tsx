import React from 'react';
import {
  styled,
  getStylePreset,
  getTypographyPreset,
  getSpacingInlineVertical,
  getMinWidth,
} from '../utils/style';
import {isValidNode} from '../utils/component';
import {OrderedListProps} from './types';
import defaults from './defaults';
import {withOwnTheme} from '../utils/with-own-theme';
import {logicalProps} from '../utils/logical-properties';

const ListItem = styled.li<OrderedListProps>`
  ${getSpacingInlineVertical('orderedList', '')}
  counter-increment: item-counter;
  ${getStylePreset('orderedList.content', 'content')}
  ${getTypographyPreset('orderedList.content', 'content', {
    withCrop: true,
  })}
  &::before {
    content: counter(item-counter) '. ';
    ${getStylePreset('orderedList.counter', 'counter')}
    ${getTypographyPreset('orderedList.counter', 'counter')}
    display: inline-block;
    min-width: ${getMinWidth('orderedList.counter', 'counter')};
  }
`;

const List = styled.ol<Pick<OrderedListProps, 'overrides'>>`
  margin: 0;
  padding: 0;
  list-style: none;
  counter-reset: item-counter;

  ${logicalProps()}
`;

const ThemelessOrderedList: React.FC<OrderedListProps> = ({
  children,
  overrides,
}) => (
  <List role="list" overrides={overrides}>
    {React.Children.map(children, node =>
      isValidNode(node) ? (
        <ListItem overrides={overrides}>{node}</ListItem>
      ) : null,
    )}
  </List>
);

export const OrderedList = withOwnTheme(ThemelessOrderedList)({defaults});
