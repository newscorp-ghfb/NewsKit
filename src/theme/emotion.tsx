import React from 'react';
import {
  withTheme as _withTheme,
  useTheme as _useTheme,
  ThemeProvider as EmotionThemeProvider,
} from '@emotion/react';
import {Theme, UncompiledTheme} from './types';
import {compileTheme} from './compiler';
import {themeDiff, CssVariablesContainer, ThemeDiff} from './css-variables';

export const withTheme = (_withTheme as unknown) as <P>(
  comp: React.ComponentType<P & {theme: Theme}>,
) => React.ComponentType<P & {theme?: Theme}>;

export const useTheme = (_useTheme as unknown) as () => Theme;

export interface ThemeProviderProps {
  theme: UncompiledTheme | Theme | ((outerTheme: Theme) => Theme);
  children: React.ReactNode;
  exposeCssVariables?: boolean;
  useThemeCache?: boolean;
}

export const ThemeProvider: React.FC<ThemeProviderProps> = ({
  theme,
  exposeCssVariables = false,
  children,
  ...props
}) => {
  const parentTheme = useTheme();
  const localTheme =
    typeof theme === 'function' ? theme(parentTheme) : compileTheme(theme);

  const diff: ThemeDiff = exposeCssVariables
    ? themeDiff(parentTheme, localTheme)
    : {};

  const doNotAddCssVariables =
    !exposeCssVariables || Object.keys(diff).length === 0;

  return (
    <EmotionThemeProvider {...props} theme={localTheme}>
      {doNotAddCssVariables ? (
        children
      ) : (
        <CssVariablesContainer diff={diff}>{children}</CssVariablesContainer>
      )}
    </EmotionThemeProvider>
  );
};
