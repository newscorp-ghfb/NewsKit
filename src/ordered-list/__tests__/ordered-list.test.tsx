import {renderToFragmentWithTheme} from '../../test/test-utils';
import {createTheme} from '../../theme';
import {OrderedList} from '../ordered-list';

const myCustomTheme = createTheme({
  name: 'my-custom-ordered-list',
  overrides: {
    stylePresets: {
      customOrderedListItemCounter: {
        base: {
          color: '{{colors.blue060}}',
        },
      },
      customOrderedListItemContent: {
        base: {
          color: '{{colors.red060}}',
        },
      },
    },
  },
});

describe('OrderedList', () => {
  test('renders with defaults', () => {
    const props = {
      children: ['first node', 'second node', 'third node'],
    };
    const fragment = renderToFragmentWithTheme(OrderedList, props);
    expect(fragment).toMatchSnapshot();
  });

  test('renders with invalid nodes', () => {
    const props = {
      children: [null, 'second node', 'third node'],
    };
    const fragment = renderToFragmentWithTheme(OrderedList, props);
    expect(fragment).toMatchSnapshot();
  });

  test('renders with overrides', () => {
    const props = {
      children: ['first node', 'second node', 'third node'],
      overrides: {
        spaceInline: 'space010',
        content: {
          stylePreset: 'customOrderedListItemContent',
          typographyPreset: 'editorialParagraph030',
        },
        counter: {
          stylePreset: 'customOrderedListItemCounter',
          typographyPreset: 'editorialParagraph020',
          minWidth: 'sizing030',
        },
      },
    };
    const fragment = renderToFragmentWithTheme(
      OrderedList,
      props,
      myCustomTheme,
    );
    expect(fragment).toMatchSnapshot();
  });

  test('renders with logical props', () => {
    const props = {
      children: ['first node', 'second node', 'third node'],
      overrides: {
        paddingInline: '30px',
        marginBlock: '30px',
      },
    };
    const fragment = renderToFragmentWithTheme(
      OrderedList,
      props,
      myCustomTheme,
    );
    expect(fragment).toMatchSnapshot();
  });
});
