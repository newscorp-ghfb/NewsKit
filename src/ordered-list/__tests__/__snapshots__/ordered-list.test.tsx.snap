// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OrderedList renders with additional HTML attributes 1`] = `
<DocumentFragment>
  .emotion-0 {
  margin: 0;
  padding: 0;
  list-style: none;
  counter-reset: item-counter;
}

<ol
    class="emotion-0"
    id="ordered-list-id"
    role="list"
  />
</DocumentFragment>
`;

exports[`OrderedList renders with defaults 1`] = `
<DocumentFragment>
  .emotion-0 {
  margin: 0;
  padding: 0;
  list-style: none;
  counter-reset: item-counter;
}

.emotion-1 {
  margin-bottom: 16px;
  counter-increment: item-counter;
  color: #3B3B3B;
  font-family: "DM Sans",sans-serif;
  font-size: 14px;
  line-height: 21px;
  font-weight: 400;
  letter-spacing: 0;
  padding: 0.5px 0px;
}

.emotion-1 svg {
  fill: #3B3B3B;
}

.emotion-1::before {
  content: '';
  margin-bottom: -0.391em;
  display: block;
}

.emotion-1::after {
  content: '';
  margin-top: -0.409em;
  display: block;
}

.emotion-1::before {
  content: counter(item-counter) '. ';
  color: #3B3B3B;
  font-family: "DM Sans",sans-serif;
  font-size: 14px;
  line-height: 1.5;
  font-weight: 400;
  letter-spacing: 0;
  display: inline-block;
  min-width: 24px;
}

.emotion-1::before svg {
  fill: #3B3B3B;
}

<ol
    class="emotion-0"
    role="list"
  >
    <li
      class="emotion-1"
    >
      first node
    </li>
    <li
      class="emotion-1"
    >
      second node
    </li>
    <li
      class="emotion-1"
    >
      third node
    </li>
  </ol>
</DocumentFragment>
`;

exports[`OrderedList renders with invalid nodes 1`] = `
<DocumentFragment>
  .emotion-0 {
  margin: 0;
  padding: 0;
  list-style: none;
  counter-reset: item-counter;
}

.emotion-1 {
  margin-bottom: 16px;
  counter-increment: item-counter;
  color: #3B3B3B;
  font-family: "DM Sans",sans-serif;
  font-size: 14px;
  line-height: 21px;
  font-weight: 400;
  letter-spacing: 0;
  padding: 0.5px 0px;
}

.emotion-1 svg {
  fill: #3B3B3B;
}

.emotion-1::before {
  content: '';
  margin-bottom: -0.391em;
  display: block;
}

.emotion-1::after {
  content: '';
  margin-top: -0.409em;
  display: block;
}

.emotion-1::before {
  content: counter(item-counter) '. ';
  color: #3B3B3B;
  font-family: "DM Sans",sans-serif;
  font-size: 14px;
  line-height: 1.5;
  font-weight: 400;
  letter-spacing: 0;
  display: inline-block;
  min-width: 24px;
}

.emotion-1::before svg {
  fill: #3B3B3B;
}

<ol
    class="emotion-0"
    role="list"
  >
    <li
      class="emotion-1"
    >
      second node
    </li>
    <li
      class="emotion-1"
    >
      third node
    </li>
  </ol>
</DocumentFragment>
`;

exports[`OrderedList renders with logical props 1`] = `
<DocumentFragment>
  .emotion-0 {
  margin: 0;
  padding: 0;
  list-style: none;
  counter-reset: item-counter;
  margin-block: 30px;
  padding-inline: 30px;
}

.emotion-1 {
  margin-bottom: 16px;
  counter-increment: item-counter;
  color: #3B3B3B;
  font-family: "DM Sans",sans-serif;
  font-size: 14px;
  line-height: 21px;
  font-weight: 400;
  letter-spacing: 0;
  padding: 0.5px 0px;
}

.emotion-1 svg {
  fill: #3B3B3B;
}

.emotion-1::before {
  content: '';
  margin-bottom: -0.391em;
  display: block;
}

.emotion-1::after {
  content: '';
  margin-top: -0.409em;
  display: block;
}

.emotion-1::before {
  content: counter(item-counter) '. ';
  color: #3B3B3B;
  font-family: "DM Sans",sans-serif;
  font-size: 14px;
  line-height: 1.5;
  font-weight: 400;
  letter-spacing: 0;
  display: inline-block;
  min-width: 24px;
}

.emotion-1::before svg {
  fill: #3B3B3B;
}

<ol
    class="emotion-0"
    role="list"
  >
    <li
      class="emotion-1"
    >
      first node
    </li>
    <li
      class="emotion-1"
    >
      second node
    </li>
    <li
      class="emotion-1"
    >
      third node
    </li>
  </ol>
</DocumentFragment>
`;

exports[`OrderedList renders with overrides 1`] = `
<DocumentFragment>
  .emotion-0 {
  margin: 0;
  padding: 0;
  list-style: none;
  counter-reset: item-counter;
}

.emotion-1 {
  margin-bottom: 4px;
  counter-increment: item-counter;
  color: #D60000;
  font-family: "DM Sans",sans-serif;
  font-size: 18px;
  line-height: 27px;
  font-weight: 400;
  letter-spacing: 0;
  padding: 0.5px 0px;
}

.emotion-1::before {
  content: '';
  margin-bottom: -0.391em;
  display: block;
}

.emotion-1::after {
  content: '';
  margin-top: -0.409em;
  display: block;
}

.emotion-1::before {
  content: counter(item-counter) '. ';
  color: #3358CC;
  font-family: "DM Sans",sans-serif;
  font-size: 16px;
  line-height: 1.5;
  font-weight: 400;
  letter-spacing: 0;
  display: inline-block;
  min-width: 12px;
}

<ol
    class="emotion-0"
    role="list"
  >
    <li
      class="emotion-1"
    >
      first node
    </li>
    <li
      class="emotion-1"
    >
      second node
    </li>
    <li
      class="emotion-1"
    >
      third node
    </li>
  </ol>
</DocumentFragment>
`;
